### PR TITLE
Router: apply select indent after a comment #1334

### LINKED ISSUE #1334: Wired alignment when chain contains comment / no parenthesis call / with parenthesis call
- **Version**: 1.5.1 (scalafmt plugin version)
- **Integration**: IntelliJ
- **Configuration**:
```
maxColumn = 150
docstrings = JavaDoc
align = none
```

## Steps

Given code like this:
```scala
class Foo {
  val vv = v.aaa //
    .aaa
    .bbb()
}
```

When I run scalafmt like this:
```bash
<press Reformat with scalafmt key>
```

## Problem

Scalafmt formats code like this:
```scala
class Foo {
  val vv = v.aaa //
  .aaa
    .bbb()
}
```

## Expectation

I would like the formatted output to look like this:
```scala
class Foo {
  val vv = v.aaa //
    .aaa
    .bbb()
}
```
Or like this:
```scala
class Foo {
  val vv = v
    .aaa //
    .aaa
    .bbb()
}
```

## Notes

I found a config key `includeNoParensInSelectChains` in the [document](https://scalameta.org/scalafmt/docs/configuration.html#includenoparensinselectchains) may related with this issue.
But after I added this key to my conf, intellij IDEA reports:
```
12:33	Scalafmt
			Failed to read /................/.scalafmt.conf. 
			Invalid fields: includeNoParensInSelectChains
```
Meanwhile the `Show original` / `Show formatted` button in the [document](https://scalameta.org/scalafmt/docs/configuration.html#includenoparensinselectchains) for this config key was not functional either at this time.

### PR DESCRIPTION
**Pull request by kitbellew**
Fixes #1334

### CODE REVIEW

**Comment by olafurpg** [comment]
Thank you for this contribution! I really appreciate you taking a look at these long standing issues 🙏 

At a quick glance this looks great but I won't be able to review properly until next week earliest. In the meantime, can you please update the PR descriptions to include before/after examples. I personally try to structure my PR descriptions in the following way

> Previously, Scalafmt formatted code like this: EXAMPLE CODE. Now, Scalafmt formats the same code like this instead: NEW FORMAT OUTPUT.

You can also include "Fixes #ISSUE" in the description so that the issue is automatically closed when merged.

File: scalafmt-tests/src/test/resources/default/Select.stat [hunk 1]
```diff
@@ -62,7 +62,7 @@ val lastToken = owner.body.tokens.filter {
   case _: Whitespace | _: Comment => false
   case _ => true
 } // edge case, if body is empty expire on arrow.
-.lastOption.getOrElse(arrow)
+  .lastOption.getOrElse(arrow)
```
Comment (poslegm) [inline scalafmt-tests/src/test/resources/default/Select.stat]: IMHO previous behavior was prettier 🤔 
@tanishiking @olafurpg what do you think?

Comment (tanishiking) [reply scalafmt-tests/src/test/resources/default/Select.stat]: I think it's acceptable to add an indent here 👍 , and it's more consistent if scalafmt adds an indent before methods for any objects (even if it's Block).

FYI [prettier add an indent in this case](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEA3TBeTYAHSk0wA8lMBycu6gGhLIE8rqXPGSBfTAen6Y0ACwgBXADYATTAHc4mAIbTZS0gEso0hFmiYYIxQDMIkyRDlaA5pkla4AfhIA6cgAoAlCAYgIABxgNdGRQJQAncMsABQiENGQQJWwIDWkfEAAjcKUwAGs4GABlf1ybZBhw8ThfERgAW0kAdRENeDRSsDgi+LaNbDaWRLA0BN8tNDhwmGic63qlZGMlSUnfACs0cgAhHPzCoqV6uAAZByWVtZBN8iKbSTgARXEIeAvVmpBS8MnwxMylJk4JIMv5wloYE00oZkAAOAAMvjBEEmTRy-kSYLgv2wcAyAEcXvBZgEEiglGgALRQOBwHTpXzhOCEjRM2ZKeaLJDLD6+Sb1DQVKqfND3J5EvHcy6fGCAqHSGFIABMvkqSg09ig1gAwhB6gtElBoHjfOJJgAVQFknlXbDVACS2l0RTA4MCAEFtEUYCwHu9JjweEA)

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 2]
```diff
@@ -1344,8 +1344,12 @@ class Router(formatOps: FormatOps) {
       case FormatToken(_, c: T.Comment, between) =>
         Seq(Split(newlines2Modification(between), 0))
       // Commented out code should stay to the left
-      case FormatToken(c: T.Comment, _, between) if c.syntax.startsWith("//") =>
-        Seq(Split(Newline, 0))
+      case FormatToken(c: T.Comment, right, _) if c.syntax.startsWith("//") =>
```
Comment (tanishiking) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: Could you add some comments that explain the behavior with some code examples? So that people who read this code can easily understand the meaning.

File: scalafmt-tests/src/test/resources/test/ContinuationIndent.stat [hunk 3]
```diff
@@ -37,3 +37,15 @@ class ExtendTest
     with D
     with E =>
 }
+<<< #1334 1
```
Comment (tanishiking) [inline scalafmt-tests/src/test/resources/test/ContinuationIndent.stat]: It might be better to move these tests to `test/resources/default/indent.stat` or something because this behavior is nothing to do with `continuationIndent`?

Comment (tanishiking) [inline scalafmt-tests/src/test/resources/test/ContinuationIndent.stat]: it'll be helpful if you name the test with more meaningful description (you can include #1334 to point the issue number of course :) )

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 2 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: will do

File: scalafmt-tests/src/test/resources/test/ContinuationIndent.stat (hunk 3 above)
Comment (kitbellew) [reply scalafmt-tests/src/test/resources/test/ContinuationIndent.stat]: no problem. perhaps i misunderstand the meaning of continuation; i thought that since we are indenting because of the `.something` which continues from previous line, it should go in this file.

Comment (kitbellew) [reply scalafmt-tests/src/test/resources/test/ContinuationIndent.stat]: that was an oversight, thanks for noticing.

File: scalafmt-tests/src/test/resources/default/Select.stat [hunk 4]
```diff
@@ -277,3 +277,78 @@ object a {
...
+<<< #1334 very long unbroken select chain
+class Foo {
+  val vv = v.aaaaaaaaaa.bbbbbbbbbb.cccccccccc.dddddddddd.eeeeeeeeee.ffffffffff.gggggggggg.hhhhhhhhhh.iiiiiiiiii
+}
+>>>
+class Foo {
+  val vv =
+    v.aaaaaaaaaa.bbbbbbbbbb.cccccccccc.dddddddddd.eeeeeeeeee.ffffffffff.gggggggggg.hhhhhhhhhh.iiiiiiiiii
+}
+<<< #1334 very long broken select chain
```
Comment (olafurpg) [inline scalafmt-tests/src/test/resources/default/Select.stat]: This change looks unrelated to comments, the current behavior of scalafmt is to format everything in a single line.

The current behavior of Scalafmt is to never break on select nodes `a.b` unless `b` is an "applied" node like `a.b()` or `a.b[T]` or `a.b { ... }`. This is an intentional decision for better or worse, and we're going to change it then we'll need careful performance testing to ensure the search state doesn't explode in cases that formatted fine before.

File: scalafmt-tests/src/test/resources/default/Select.stat (hunk 1 above)
Comment (olafurpg) [reply scalafmt-tests/src/test/resources/default/Select.stat]: I like the new behavior.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 5]
```diff
@@ -1340,6 +1340,16 @@ class Router(formatOps: FormatOps) {
           Split(Space, 0, policy = singleLine),
           Split(Newline, 1).withPolicy(penalizeNewlineByNesting(cond, arrow))
         )
+      case FormatToken(_, dot @ T.Dot(), _) =>
+        // allow keeping a newline (and indent) within a select chain
+        val noNewLine = !style.activeForEdition_2019_11 ||
+          newlines == 0 || rightOwner.isNot[Term.Select]
+        val newlinePenalty = if (noNewLine) 0 else 10 + treeDepth(rightOwner)
+        Seq(
```
Comment (olafurpg) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: FWIW, doubling the search state here is likely to introduce new "search state exploded" errors for programs that formatted fine before.
